benchmark/benchmark.php

```diff
@@ -98,41 +100,91 @@ function runValgrindPhpCgiCommand(
     array $args,
     ?string $cwd = null,
     bool $jit = false,
-    int $warmup = 0,
     int $repeat = 1,
 ): array {
     global $phpCgi;
 
-    $profileOut = __DIR__ . "/profiles/callgrind.out.$name";
+    $profileOut = __DIR__ . "/profiles/callgrind.$name";
     if ($jit) {
-        $profileOut .= '.jit';
+        $profileOut .= '-jit';
     }
 
     $process = runCommand([
         'valgrind',
         '--tool=callgrind',
         '--dump-instr=yes',
+        '--collect-jumps=yes',
+        '--cache-sim=yes',
+        '--branch-sim=yes',
         "--callgrind-out-file=$profileOut",
+        '--verbose',
         '--',
         $phpCgi,
-        '-T' . ($warmup ? $warmup . ',' : '') . $repeat,
+        '-T' . $repeat,
         '-d max_execution_time=0',
         '-d opcache.enable=1',
         '-d opcache.jit=' . ($jit ? 'tracing' : 'disable'),
         '-d opcache.jit_buffer_size=128M',
         '-d opcache.validate_timestamps=0',
         ...$args,
     ]);
-    $instructions = extractInstructionsFromValgrindOutput($process->stderr);
-    if ($repeat > 1) {
-        $instructions = gmp_strval(gmp_div_q($instructions, $repeat));
+
+    // collect metrics for startup, each benchmark run and shutdown
+    $metricsArr = [];
+    foreach (['startup' => 1, ...range(2, $repeat + 1), 'shutdown' => ''] as $k => $kCallgrindOut) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+        rename($profileOut . ($kCallgrindOut === '' ? '' : '.' . $kCallgrindOut), $profileOutSpecific);
+
+        $metricsArr[$k] = extractMetricsFromCallgrindFile($profileOutSpecific);
     }
-    return ['instructions' => $instructions];
+
+    // print all collected metrics
+    print_r(array_map(fn ($metrics) => array_map(fn ($v) => number_format($v, 0, '.', '_'), $metrics), $metricsArr));
+
+    // find the fastest benchmark run
+    $bestRunIndex = 0;
+    foreach (range(0, $repeat - 1) as $k) {
+        if ($metricsArr[$k]['Ir'] < $metricsArr[$bestRunIndex]['Ir']) {
+            $bestRunIndex = $k;
+        }
+    }
+
+    // remove non-fastest profiles from artifacts
+    foreach (range(0, $repeat - 1) as $k) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+
+        if ($k !== $bestRunIndex) {
+            unlink($profileOutSpecific);
+        }
+    }
+
+    // annotate profiles for artifacts
+    foreach (['startup', $bestRunIndex, 'shutdown'] as $k) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+
+        runCommand([
+            'callgrind_annotate',
+            '--threshold=100',
+            '--auto=yes',
+            '--show-percs=no',
+            $profileOutSpecific,
+            new UnescapedArg('>'),
+            "$profileOutSpecific.txt",
+        ]);
+    }
+
+    return ['instructions' => $metricsArr[$bestRunIndex]['Ir']];
```

Comment:
```suggestion
    return $metricsArr;
```

should be returned and stored in historical data